src/Bicep.Core/TypeSystem/TypeValidator.cs

```diff
@@ -177,16 +177,31 @@ public static TypeSymbol NarrowTypeAndCollectDiagnostics(ITypeManager typeManage
             }
 
             // object assignability check
-            if (expression is ObjectSyntax objectValue && targetType is ObjectType targetObjectType)
+            if (expression is ObjectSyntax objectValue)
             {
-                return NarrowObjectType(typeManager, objectValue, targetObjectType, diagnosticWriter, skipConstantCheck);
+                switch (targetType)
+                {
+                    case ObjectType targetObjectType:
+                        return NarrowObjectType(typeManager, objectValue, targetObjectType, diagnosticWriter, skipConstantCheck);
+
+                    case DiscriminatedObjectType targetDiscriminated:
+                        return NarrowDiscriminatedObjectType(typeManager, objectValue, targetDiscriminated, diagnosticWriter, skipConstantCheck);
+                }
             }
 
-            if (expression is ObjectSyntax objectDiscriminated && targetType is DiscriminatedObjectType targetDiscriminated)
+            // if-condition assignability check
+            if (expression is IfConditionSyntax {Body:ObjectSyntax body})

```

Comment:
`ForSyntax` will have the same issue here, so I'll probably introduce some new interface.